src/lib.rs

```diff
@@ -750,12 +751,67 @@ pub trait Rng: RngCore + Sized {
     /// println!("{:?}", y);
     /// ```
     fn shuffle<T>(&mut self, values: &mut [T]) {
+        use distributions::range::WideningMultiply;
+        // In theory this function is nothing more then:
+        // ```
+        // while i >= 2 {
+        //     i -= 1;
+        //     values.swap(i, self.gen_range(0, i + 1));`
+        // }
+        // ```
+        //
+        // `gen_range` samples completely unbiased from the range. It can be 4x
+        // as fast if only we could sample with a tiny bias. Using a modulus to
+        // reduce the range will always show a bias towards the lower numbers
+        // in the range.
+        //
+        // But we use the widening multiply method. Here the bias is evenly
+        // distributed over all numbers in the range. Also for every
+        // iteration the range is 1 less then the previous iteration, and
+        // the numbers that might have a bias are different. So besides the
+        // biases being very tiny, they also partly cancel out each other.
+        //
+        // Said differently, we only have to worry about biases when a whole
+        // group together shows a bias, as in the modulus method.
+
+        // We also optimize for slices of different length: huge slices with
+        // more than 2^32 elements (usize indexes), large slices with more than
+        // 2^16 elements, and medium to small slices. This allows us to make
+        // better use of the bits generated by the RNG.
         let mut i = values.len();
-        while i >= 2 {
-            // invariant: elements with index >= i have been locked in place.
+        while i > core::u32::MAX as usize || i & 1 == 1 {
+            // Invariant: elements with index >= i have been locked in place.
+            i -= 1;
+            // Lock element i in place, by swapping it with a random element in
+            // the range [0, i] (inclusive).
+            let (j, _) = self.gen::<u64>().wmul((i + 1) as u64);
+            values.swap(i, j as usize);
+        }
+        let mut i = i as u32;
+
+        while i > core::u16::MAX as u32 || i & 2 == 2 {
+            let r = self.gen::<u64>();
+            let r1 = r as u32;
+            let r2 = (r >> 32) as u32;
+
             i -= 1;
-            // lock element i in place.
-            values.swap(i, self.gen_range(0, i + 1));
+            let (j, _) = r1.wmul(i + 1);
+            values.swap(i as usize, j as usize);
+
+            i -= 1;
+            let (j, _) = r2.wmul(i + 1);
+            values.swap(i as usize, j as usize);
+        }
+
+        let mut i = i as u16;
+        while i >= 2 {
```

Comment:
Shouldn't this be `while i >= 5` with special cases for the last choices? This algorithm could wrap `i`. Also implies the guards on above while loops may need changing.

Reply:
That is also possible. My thought was to handle the special cases before starting a loop. Maybe doesn't matter anything.